src/internal_rules.jl

```diff
@@ -558,3 +558,244 @@ function EnzymeRules.reverse(
     xs.dval .= xs.dval[back_inds]
     return (nothing,)
 end
+
+function EnzymeRules.forward(
+    ::Const{typeof(cholesky)},
+    RT::Type{<:Union{Const, DuplicatedNoNeed, Duplicated}},
+    A::Union{Const, Duplicated};
+    kwargs...
+)
+    fact = cholesky(A.val; kwargs...)
+    if RT <: Const
+        return fact
+    end
+    # TODO: This will be a problem for sparse matrices as invL and dL are dense
+    invL = inv(fact.L)
+    # TODO: dL is dense even when L was sparse
+    dL = Matrix(fact.L * LowerTriangular(invL * A.dval * invL' * 0.5 * I))
+    # TODO: Stored as Cholesky, although it isn't a Cholesky factorization
+    dfact = Cholesky(dL, 'L', 0)
+    if RT <: DuplicatedNoNeed
+        return dfact
+    else
+        return Duplicated(fact, dfact)
+    end
+end
+
+function EnzymeRules.forward(
+    ::Const{typeof(cholesky)},
+    RT::Type{<:Union{BatchDuplicatedNoNeed, BatchDuplicated}},
+    A::Union{BatchDuplicatedNoNeed{T,N}, BatchDuplicated{T,N}};
+    kwargs...
+) where {T,N}
+    fact = cholesky(A.val; kwargs...)
+    if RT <: Const
+        return fact
+    end
+    invL = inv(fact.L)
+    dfact = ntuple(
+        i-> Cholesky(
+            Matrix(fact.L * LowerTriangular(invL * A.dval[i] * invL' * 0.5 * I)), 'L', 0
+        ), Val(N)
+    )
+    if RT <: BatchDuplicatedNoNeed
+        return dfact
+    else
+        return BatchDuplicated(fact, dfact)
+    end
+end
+
+function EnzymeRules.forward(
+        ::Const{typeof(\)},
+        RT::Type{<:Union{Const, DuplicatedNoNeed, Duplicated}},
+        fact::Union{Const, Duplicated}, B::Union{Const, Duplicated};
+        kwargs...
+)
+    retval = copy(B.val)
+    ldiv!(fact.val, retval)
+    retdval = []
+    if isa(fact, Duplicated) && isa(B, Duplicated)
+        retdval = zeros(length(retval))
+        mul!(retdval, fact.dval.U, retval)
+        mul!(retdval, fact.dval.L, retdval)
+        retdval .= B.dval .- retdval
+        ldiv!(fact.val, retdval)
+    elseif isa(fact, Duplicated) && isa(B, Const)
+        retdval = zeros(length(retval))
+        mul!(retdval, A, B.val)
+        mul!(retdval, -1, retdval)
+        ldiv!(fact.val, retdval)
+    elseif isa(fact, Const) && isa(B, Duplicated)
+        retdval = copy(B.dval)
+        ldiv!(fact.val, retdval)
+    elseif isa(fact, Const) && isa(B, Const)
+        nothing
+    else
+        error("Error in forward \\ Enzyme rule $(typeof(fact)) $(typeof(B)).")
+    end
+    if RT <: Const
+        return retval
+    elseif (RT <: DuplicatedNoNeed) || (RT <: BatchDuplicatedNoNeed)
+        return retdval
+    else
+        return Duplicated(retval, retdval)
+    end
+end
+
+function EnzymeRules.forward(
+        ::Const{typeof(\)},
+        RT::Type{<:Union{BatchDuplicatedNoNeed, BatchDuplicated}},
+        fact::Union{Const, BatchDuplicated{T1,N}}, B::Union{Const, BatchDuplicated{T2,N}};
+        kwargs...
+) where {T1,T2,N}
+    retval = copy(B.val)
+    ldiv!(fact.val, retval)
+    bretdval = ntuple(Val(N)) do b
```

Comment:
This should be parallelized somehow. I don't know how.

Reply:
For now tuple is fine and standard practice